frontend/vue/components/CodeExercise/CodeExercise.vue

```diff
@@ -0,0 +1,153 @@
+<template>
+  <div class="code-exercise">
+    <div class="code-exercise__editor-block">
+      <CodeEditor
+        class="code-exercise__editor-block__editor"
+        :code="code"
+        :initial-code="initialCode"
+        :notebook-enabled="false"
+        @codeChanged="codeChanged"
+        @notebookCopyRequest="notebookCopyRequest"
+      />
+      <ExerciseActionsBar
+        class="code-exercise__editor-block__actions-bar"
+        :is-running="isKernelBusy"
+        :run-enabled="isKernelReady"
+        :grade-enabled="isKernelReady && isGradingExercise"
+        @run="run"
+        @grade="grade"
+      />
+    </div>
+    <CodeOutput
+      ref="output"
+      class="code-exercise__output"
+      @running="kernelRunning"
+      @finished="kernelFinished"
+      @kernelReady="kernelReady"
+    />
+    <div
+      ref="initialCodeElement"
+      class="code-exercise__initial-code"
+      :class="{'code-exercise__initial-code__hidden-output': executedOnce}"
+    >
+      <slot />
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+import { defineComponent } from 'vue-demi'
+import CodeEditor from './CodeEditor.vue'
+import ExerciseActionsBar from './ExerciseActionsBar.vue'
+import CodeOutput from './CodeOutput.vue'
+
+export default defineComponent({
+  name: 'CodeExercise',
+  components: {
+    CodeEditor,
+    ExerciseActionsBar,
+    CodeOutput
+  },
+  props: {
+    graderImport: {
+      type: String,
+      required: false,
+      default: ''
+    },
+    graderFunction: {
+      type: String,
+      required: false,
+      default: ''
+    }
+  },
+  data () {
+    return {
+      code: '',
+      initialCode: '',
+      isKernelBusy: false,
+      isKernelReady: false,
+      executedOnce: false
+    }
+  },
+  computed: {
+    isGradingExercise (): boolean {
+      return this.graderFunction !== '' && this.graderImport !== ''
+    }
+  },
+  mounted () {
+    const slotWrapper = (this.$refs.initialCodeElement as HTMLDivElement)
+    const initialCodeElement = slotWrapper.getElementsByTagName('pre')[0]
+    this.code = initialCodeElement?.textContent?.trim() ?? ''
+    this.initialCode = this.code
+  },
+  methods: {
+    run () {
+      const codeOutput: any = this.$refs.output
+      codeOutput.requestExecute(this.code)
+    },
+    grade () {
+      /* TBD */
```

Comment:
Prefer commenting as "TODO" to find it easier.

```suggestion
      /* TODO */
```